api/services/syncing/syncing.go

```diff
@@ -260,78 +302,267 @@ func (ss *StateSync) GetConsensusHashes() bool {
+	ss.syncMux.Lock()
+	worker.UpdateCurrent()
+	ss.syncMux.Unlock()
+	return true
+}
+
+// generateNewState will construct most recent state from downloaded blocks
+func (ss *StateSync) generateNewState(bc *core.BlockChain, worker *worker.Worker) {
+	// update blocks created before node start sync
+	parentHash := bc.CurrentBlock().Hash()
+	for {
+		block := ss.getBlockFromOldBlocksByParentHash(parentHash)
+		if block == nil {
+			break
+		}
+		ok := ss.updateBlockAndStatus(block, bc, worker)
+		if !ok {
+			break
+		}
+		parentHash = block.Hash()
+	}
+	ss.syncMux.Lock()
+	ss.commonBlocks = make(map[int]*types.Block)
+	ss.syncMux.Unlock()
+
+	// update blocks after node start sync
+	parentHash = bc.CurrentBlock().Hash()
+	for {
+		block := ss.getMaxConsensusBlockFromParentHash(parentHash)
+		if block == nil {
+			break
+		}
+		ok := ss.updateBlockAndStatus(block, bc, worker)
+		if !ok {
+			break
+		}
+		parentHash = block.Hash()
+	}
+	ss.syncMux.Lock()
+	for id := range ss.syncConfig.peers {
+		ss.syncConfig.peers[id].newBlocks = []*types.Block{}
+	}
+	ss.syncMux.Unlock()
+
+	// update last mile blocks if any
+	parentHash = bc.CurrentBlock().Hash()
+	for {
+		block := ss.getBlockFromLastMileBlocksByParentHash(parentHash)
+		if block == nil {
+			break
+		}
+		ok := ss.updateBlockAndStatus(block, bc, worker)
+		if !ok {
+			break
+		}
+		parentHash = block.Hash()
+	}
 }
 
 // StartStateSync starts state sync.
-func (ss *StateSync) StartStateSync(peers []p2p.Peer, bc *core.BlockChain) bool {
-	// Creates sync config.
-	ss.CreateSyncConfig(peers)
-	// Makes connections to peers.
-	ss.MakeConnectionToPeers()
+func (ss *StateSync) StartStateSync(startHash []byte, bc *core.BlockChain, worker *worker.Worker) {
+	ss.RegisterNodeInfo()
 	// Gets consensus hashes.
-	if !ss.GetConsensusHashes() {
-		return false
+	if !ss.GetConsensusHashes(startHash) {
+		Log.Debug("[sync] StartStateSync unable to reach consensus on ss.GetConsensusHashes")
+		return
 	}
+	Log.Debug("[sync] StartStateSync reach consensus on ss.GetConsensusHashes")
 	ss.generateStateSyncTaskQueue(bc)
 	// Download blocks.
 	if ss.stateSyncTaskQueue.Len() > 0 {
 		ss.downloadBlocks(bc)
 	}
-	return true
+	ss.generateNewState(bc, worker)
+}
+
+func (peerConfig *SyncPeerConfig) registerToBroadcast(peerHash []byte) error {
+	response := peerConfig.client.Register(peerHash)
+	if response == nil || response.Type == pb.DownloaderResponse_FAIL {
+		return ErrRegistrationFail
+	} else if response.Type == pb.DownloaderResponse_SUCCESS {
+		return nil
+	}
+	return ErrRegistrationFail
+}
+
+// RegisterNodeInfo will register node to peers to accept future new block broadcasting
+// return number of successfull registration
+func (ss *StateSync) RegisterNodeInfo() int {
+	ss.CleanUpNilPeers()
+	registrationNumber := RegistrationNumber
+	Log.Debug("[sync]", "registrationNumber", registrationNumber, "activePeerNumber", ss.activePeerNumber)
```

Comment:
any statement after [sync], it is unclear.

Reply:
add some comments